pkg/restapi/operation/operations.go

```diff
@@ -0,0 +1,340 @@
+}
+
+type idToken interface {
+	Claims(interface{}) error
+}
+
+type oauth2Config interface {
+	AuthCodeURL(string, ...oauth2.AuthCodeOption) string
+	Exchange(context.Context, string, ...oauth2.AuthCodeOption) (oauth2Token, error)
+}
+
+type oauth2ConfigImpl struct {
+	oc *oauth2.Config
+}
+
+func (o *oauth2ConfigImpl) AuthCodeURL(state string, options ...oauth2.AuthCodeOption) string {
+	return o.oc.AuthCodeURL(state, options...)
+}
+
+func (o *oauth2ConfigImpl) Exchange(
+	ctx context.Context, code string, options ...oauth2.AuthCodeOption) (oauth2Token, error) {
+	return o.oc.Exchange(ctx, code, options...)
+}
+
+type oauth2Token interface {
+	Extra(string) interface{}
+}
+
+// Operation defines handlers.
+type Operation struct {
+	handlers         []Handler
+	client           httpClient
+	requestTokens    map[string]string
+	transientStore   storage.Store
+	oidcProvider     oidcProvider
+	oidcClientID     string
+	oidcClientSecret string
+	oidcCallbackURL  string
+	oauth2ConfigFunc func(...string) oauth2Config
+}
+
+// Config defines configuration for rp operations.
+type Config struct {
+	TLSConfig              *tls.Config
+	RequestTokens          map[string]string
+	OIDCProviderURL        string
+	OIDCClientID           string
+	OIDCClientSecret       string
+	OIDCCallbackURL        string
+	TransientStoreProvider storage.Provider
+}
+
+type createOIDCRequestResponse struct {
+	Request string `json:"request"`
+}
+
+// New returns rp operation instance.
+func New(config *Config) (*Operation, error) {
+	svc := &Operation{
+		client:           &http.Client{Transport: &http.Transport{TLSClientConfig: config.TLSConfig}},
+		requestTokens:    config.RequestTokens,
+		oidcClientID:     config.OIDCClientID,
+		oidcClientSecret: config.OIDCClientSecret,
+		oidcCallbackURL:  config.OIDCCallbackURL,
+	}
+
+	idp, err := oidc.NewProvider(
+		oidc.ClientContext(
+			context.Background(),
+			&http.Client{
+				Transport: &http.Transport{TLSClientConfig: config.TLSConfig},
+			},
+		),
+		config.OIDCProviderURL,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to init oidc provider with url [%s] : %w", config.OIDCProviderURL, err)
+	}
+
+	svc.oidcProvider = &oidcProviderImpl{op: idp}
+
+	svc.transientStore, err = createStore(config.TransientStoreProvider)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create store : %w", err)
+	}
+
+	svc.oauth2ConfigFunc = func(scopes ...string) oauth2Config {
+		config := &oauth2.Config{
+			ClientID:     svc.oidcClientID,
+			ClientSecret: svc.oidcClientSecret,
+			Endpoint:     svc.oidcProvider.Endpoint(),
+			RedirectURL:  fmt.Sprintf("%s%s", svc.oidcCallbackURL, oauth2CallbackPath),
+			Scopes:       []string{oidc.ScopeOpenID},
+		}
+
+		if len(scopes) > 0 {
+			config.Scopes = append(config.Scopes, scopes...)
+		}
+
+		return &oauth2ConfigImpl{oc: config}
+	}
+
+	svc.registerHandler()
+
+	return svc, nil
+}
+
+func (c *Operation) createOIDCRequest(w http.ResponseWriter, r *http.Request) {
```

Comment:
note: the OIDC request needs to be tailored to the OIDC provider

Reply:
https://github.com/trustbloc/hub-auth/issues/9 that will be my next task. OIDC requests specifically tailored for google provider and similarly for the others eventually.